clustering/common/src/main/java/org/jboss/as/clustering/controller/RestartParentAddHandler.java

```diff
@@ -46,33 +40,28 @@
  */
 public class RestartParentAddHandler<T> extends RestartParentResourceAddHandler implements Registration {
 
-    private final ResourceDescriptionResolver resolver;
+    private final AddStepHandlerDescriptor descriptor;
     private final ResourceServiceBuilderFactory<T> builderFactory;
-    private final List<Attribute> attributes = new LinkedList<>();
 
-    public RestartParentAddHandler(ResourceDescriptionResolver resolver, ResourceServiceBuilderFactory<T> builderFactory) {
+    public RestartParentAddHandler(AddStepHandlerDescriptor descriptor, ResourceServiceBuilderFactory<T> builderFactory) {
         super(null);
-        this.resolver = resolver;
+        this.descriptor = descriptor;
         this.builderFactory = builderFactory;
     }
 
-    public <E extends Enum<E> & Attribute> RestartParentAddHandler<T> addAttributes(Class<E> enumClass) {
-        return this.addAttributes(EnumSet.allOf(enumClass));
-    }
-
-    public RestartParentAddHandler<T> addAttributes(Attribute... attributes) {
-        return this.addAttributes(Arrays.asList(attributes));
-    }
-
-    public RestartParentAddHandler<T> addAttributes(Collection<? extends Attribute> attributes) {
-        this.attributes.addAll(attributes);
-        return this;
+    @Override
+    protected void updateModel(OperationContext context, ModelNode operation) throws OperationFailedException {
+        super.updateModel(context, operation);
+        PathAddress address = context.getCurrentAddress();
+        for (Capability capability : this.descriptor.getCapabilities()) {
+            context.registerCapability(capability.getRuntimeCapability(address), null);
+        }
```

Comment:
There's an assumption here that all capabilities will be dynamic. This currently may be true, but something to consider.